src/util/mod.rs

```diff
@@ -0,0 +1,149 @@
+use rand::prelude::*;
+use std::collections::BTreeMap;
+use std::collections::HashMap;
+use std::collections::VecDeque;
+use std::time::{SystemTime, UNIX_EPOCH};
+
+const DEFAULT_BACKOFF_TIME_MAX_MS: f64 = 24f64 * 60f64 * 60f64 * 1000f64;
+const DEFAULT_BACKOFF_POW: f64 = 1.05f64;
+const DEFAULT_BACKOFF_TIME_MS: f64 = 5f64 * 1000f64;
+
+#[napi]
+
+pub struct QueueWithBackoff {
+    queue: VecDeque<String>,
+    /**
+     * A map of absolute backoff timestamps mapped to the value.
+     */
+    backoff: BTreeMap<u64, String>,
+    /**
+     * The last duration applied when a value was backed off.
+     */
+    last_backoff_duration: HashMap<String, u32>,
+    backoff_time: f64,
+    backoff_exponent: f64,
+    backoff_max: f64,
+}
+
+impl Default for QueueWithBackoff {
+    fn default() -> Self {
+        Self::new(
+            DEFAULT_BACKOFF_TIME_MS,
+            DEFAULT_BACKOFF_POW,
+            DEFAULT_BACKOFF_TIME_MAX_MS,
+        )
+    }
+}
+#[napi]
+
+impl QueueWithBackoff {
+    #[napi(constructor)]
+    pub fn new(backoff_time: f64, backoff_exponent: f64, backoff_max: f64) -> Self {
+        QueueWithBackoff {
+            queue: VecDeque::new(),
+            backoff: BTreeMap::new(),
+            last_backoff_duration: HashMap::new(),
+            backoff_time,
+            backoff_exponent,
+            backoff_max,
+        }
+    }
+
+    #[napi]
+    pub fn pop(&mut self) -> Option<String> {
+        let start = SystemTime::now();
+        let since_the_epoch = start.duration_since(UNIX_EPOCH).unwrap().as_millis() as u64;
+
+        // We only need to check this once, as we won't be adding to the backoff queue
+        // as often as we pull from it.
+        if let Some(item) = self.backoff.first_entry() {
+            if *item.key() < since_the_epoch {
+                let v = item.remove();
+                self.queue.push_back(v);
+            }
+        }
```

Comment:
Could you explain this part?  I'm reading it as "if the lowest backoff duration has passed, enqueue the URL that the backoff was for".  If that's the case, why is it safe to check that only once?  Is it because the queue is only for items whose backoff has expired & are thus always fair game to be fetched -- and because if no item has finished its backoff, this will return `None` and the caller will see no feeds available to poll?

Reply:
Sure, and please check my math here. We check the backoff queue once per `pull` call, which means we will pull one item (if available) and push it onto the "active" queue.

Each time we pull, we will do an action and then either push onto the active queue or the backoff queue. This means at worst case, we are only ever putting onto the backoff queue as often as we are pulling from it. In reality, it's unlikely that you are backing off every feed but if you were, you'd essentially just have a queue with a delay in it. The other thing to note here is that backed off items are inherently delayed, so that it's reasonably possible that we are `pull`ing more often than the items become "ready".

Now, it is possible that several items could all become ready within a few `ms` and the pull code isn't called exactly when they become ready. That's not really a problem, because we don't mind if items that are already failing are waiting a bit longer -- as long as they eventually get polled, they will enter the fast lane again. 

So to summarise why this is safe, we won't ever push to the backoff queue more often than we pull.

This is also efficient, because we're doing less work per pull (not iterating) and end users should notice no difference.

Does that make more sense?

Reply:
Indeed it does, thanks!